server/aesdsocket.c

```diff
@@ -0,0 +1,377 @@
+/**
+ * @file    aesdsocket.c
+ * @brief   This program creates a socket server that listens on port 9000, accepts incoming connections,
+ *          and receives data from clients. It stores the received data in a file at "/var/tmp/aesdsocketdata"
+ *          and sends back the accumulated data to the clients.
+ *			
+ * @date    October 5, 2023
+ * @author  Rishikesh Sundaragiri
+ */
+#include <stdio.h>
+#include <stdbool.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h> //has the struct addrinfo variable here
+#include <syslog.h>
+#include <unistd.h>
+#include <signal.h>
+#include <sys/stat.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <sys/wait.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <getopt.h>
+#include <errno.h>
+
+#define PORT "9000"
+#define BACKLOG 10
+#define ERROR_CODE -1
+#define DATA_FILE "/var/tmp/aesdsocketdata"
+
+
+int main_sockfd = 0;
+int client_sockfd = 0;
+int data_file_fd = 0;
+
+/*
+ * Reference : https://beej.us/guide/bgnet/html/#cb47-58
+*/
+void *get_in_addr(struct sockaddr *sa)
+{
+    if (sa->sa_family == AF_INET) {
+        return &(((struct sockaddr_in*)sa)->sin_addr);
+    }
+
+    return &(((struct sockaddr_in6*)sa)->sin6_addr);
+}
+
+/*
+ *  Executes whenever a  SIGINT or SIGTERM is received
+ *  parameter: int sig is the received signal number
+ *  NOTE : My initial idea was to keep the signal handler as short as possible. So I just set a flag here and used to ckeck that flag inside while(1)
+ *  to detech a signal occurance. But by doing that the Crtl+C command was not working. So I had to put the entire logic of cleanup inside the signal handler 
+ *  to make it work. But I still hesitate to put the cleanup code here because its not safe.
+ */
+void sig_handler(int sig)
+{
+    if(-1 == close(main_sockfd))
+    {
+        perror("close(main_sockfd)");
+        exit(ERROR_CODE);  
+    }
+    if(-1 == close(data_file_fd))
+    {
+        perror("close(data_file_fd)");
+        exit(ERROR_CODE);                 
+    }
+
+    syslog(LOG_DEBUG,"Caught signal, exiting");
+    closelog();
+
+    if(-1 == remove(DATA_FILE))
+    {
+        perror("remove(DATA_FILE)");
+    }
```

Comment:
@Rishikesh38 as you mentioned in your comment, not all of this is safe to do in signal handler.  It would be better to implement in a different way.  We will discuss in class.